drivers/spi/spi_nrfx_spim.c

```diff
@@ -378,17 +379,30 @@ static int transceive(const struct device *dev,
 
 	spi_context_lock(&dev_data->ctx, asynchronous, cb, userdata, spi_cfg);
 
+	(void)pm_device_runtime_get(dev);
+
 	error = configure(dev, spi_cfg);
-	if (error == 0) {
-		dev_data->busy = true;
+	if (error < 0) {
+		goto release;
+	}
 
-		spi_context_buffers_setup(&dev_data->ctx, tx_bufs, rx_bufs, 1);
-		spi_context_cs_control(&dev_data->ctx, true);
+	dev_data->busy = true;
 
-		transfer_next_chunk(dev);
+	spi_context_buffers_setup(&dev_data->ctx, tx_bufs, rx_bufs, 1);
+	spi_context_cs_control(&dev_data->ctx, true);
 
-		error = spi_context_wait_for_completion(&dev_data->ctx);
-	}
+	transfer_next_chunk(dev);
+
+	error = spi_context_wait_for_completion(&dev_data->ctx);
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	/* TODO: remove once nrfx automatically disables the device */
+	nrfx_spim_uninit(&dev_config->spim);
+	dev_data->initialized = false;
+#endif
+
+release:
+	(void)pm_device_runtime_put(dev);
```

Comment:
btw, I realized we can't do that here in case of async transfers. @anangl could you suggest a workflow?

Reply:
You can check the `ctx->asynchronous` flag here and perform required operations for synchronous transfers only. But for asynchronous transfers, I don't see an easy way of handling this. It looks like a workqueue would need to be involved.

Reply:
we actually have the put_async variant.

Reply:
Yes. but AFAICT, it also cannot be called from the ISR context.

Reply:
It should be safe, since it just calls `k_work_schedule` with K_NO_WAIT underneath.

Reply:
And isn't this problematic? https://github.com/zephyrproject-rtos/zephyr/blob/8a6642af8a83449d6631c21bfa0f0c15de46d47b/subsys/pm/device_runtime.c#L47.

Reply:
indeed, it is. we should really add K_NO_WAIT in the async path, or at least when in ISR context. Let me fix that.